python/ray/gateway.py

```diff
@@ -0,0 +1,110 @@
+import click
+from flask import Flask, request, send_file
+import io
+import pickle
+import pyarrow
+from pyarrow import plasma as plasma
+import sys
+
+app = Flask(__name__)
+ctx = pyarrow.default_serialization_context()
+
+@app.route('/', methods=['GET', 'POST'])
+def index():
+    """TODO (dsuo): Add comments.
+
+    TODO (dsuo): Maybe care about batching
+    TODO (dsuo): REST is nice conceptually, but probably too much overhead
+    TODO (dsuo): Move logic to C++ (keep as head node process)
+
+    NOTE: We do an extra serialize during get and extra deserialize during
+        put.
+    """
+    if request.method == 'POST':
+        object_id = request.files['object_id'].read()
+        data = request.files['value'].read()
+        value = ctx.deserialize(data)
+
+        plasma_client.put(
```

Comment:
We don't necessarily need to address this right now, but this can raise an exception (e.g., if the store is full). We probably don't want to let that bring down the gateway. Do we want to propagate the exception back to the `Client`?

Reply:
Hadn't thought about this, but in general we do want to propagate some sort of logging / exceptions back to the client. 

We could tackle exceptions first. Is there a ```ray``` way that the team has settled on? There are a bunch of ways (e.g., return codes, pickled exception objects).

I imagine logging could work somewhat like how workers communicate logs back to the driver now.

Reply:
There are a couple different exception mechanisms.

1. If a task raises an exception, then we essentially store a custom exception object in the object store. When a worker/driver retrieves that object, it checks the type and will raise an exception.
2. If some issue occurs in the backend or somewhere other than in a task, and we want to bring that issue to the attention of the driver, we have a mechanism for that. From C++ it is https://github.com/ray-project/ray/blob/7edc677304f90bcab0a01e9722e59e36dd9afcb0/src/ray/gcs/tables.h#L470-L471 and from Python it is https://github.com/ray-project/ray/blob/7edc677304f90bcab0a01e9722e59e36dd9afcb0/python/ray/utils.py#L53-L57 (which uses the same C++ code path).

We'll need something similar to mechanism 1 for the gateway. That is
- We need a normal way to return values from the gateway to external client.
- For tasks that raise exceptions, this return value should just be a special exception object.
- When the external client gets a return value, it checks to see if the object is a special exception object and if so, it raises an exception.

What do you think about something like that?

Note that this doesn't need to be part of the first pass. We can always leave a TODO.